### PR TITLE
[enhancement] analyze filepath via globby

### PR DESCRIPTION
**Pull request by Genuifx**
learn from the @NumerHero, I use globby to analyze filepath~
image

### CODE REVIEW

**Comment by Genuifx** [comment]
I already install dependencies via npm, also lint code with eslint --fix. what wrong now?

**Comment by NumerHero** [comment]
thx for your pay out

**Comment by SimenB** [comment]
@Genuifx You can ignore the hound failures, they have some issues with eslint 4

File: src/handleUrl.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import globby from 'globby';
+
+export default async function(assets) {
+  const globbyAssets = [];
+  const normalAssets = [];
+  try {
```
Comment (SimenB) [inline src/handleUrl.js]: Why is this `try` necessary?

File: src/handleUrl.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        globby.hasMagic(asset.filepath)
+          ? globbyAssets.push(asset)
+          : normalAssets.push(asset)
+    );
+  } catch (e) {
+    return assets;
+  }
+  const ret = [];
+  const promises = [];
+  for (let i = 0; i < globbyAssets.length; i++) {
```
Comment (SimenB) [inline src/handleUrl.js]: use `globbyAssets.map` instead?

File: test.js [hunk 3]
```diff
@@ -295,3 +296,15 @@ test('filter option should include some files with string option', async () => {
   expect(callback).toHaveBeenCalledTimes(1);
   expect(callback).toHaveBeenCalledWith(null, pluginData);
 });
+
+test('use globby to find multi file', async () => {
+  const assets = [{ filepath: './src/*.js' }];
+  const ret = await handleUrl(assets);
+  expect(ret.length).toBeGreaterThanOrEqual(1);
```
Comment (SimenB) [inline test.js]: `expect(ret).toHaveLength(2)` or whatever the length is

File: src/handleUrl.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import globby from 'globby';
+
+export default async function(assets) {
```
Comment (SimenB) [inline src/handleUrl.js]: Could you add a comment above with what this functions does?

File: test.js [hunk 5]
```diff
@@ -295,3 +296,15 @@ test('filter option should include some files with string option', async () => {
...
+});
+
+test('filepath without globbyMagic should just return', async () => {
+  const assets = [{ filepath: path.join(__dirname, 'my-file.js') }];
+  const ret = await handleUrl(assets);
+  expect(ret.length).toBe(1);
```
Comment (SimenB) [inline test.js]: `expect(ret).toHaveLength(1)`

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+    "eslint.enable": true
+}
```
Comment (SimenB) [inline .vscode/settings.json]: Can you add a trailing newline?

File: example/webpack.config.js [hunk 7]
```diff
@@ -20,7 +20,7 @@ module.exports = {
     }),
     new HtmlWebpackPlugin(),
     new AddAssetHtmlPlugin({
-      filepath: require.resolve('./build/vendor.dll.js'),
+      filepath: path.resolve(__dirname, './build/*.dll.js'),
```
Comment (SimenB) [inline example/webpack.config.js]: why is this necessary?

File: src/handleUrl.js (hunk 1 above)
Comment (Genuifx) [inline src/handleUrl.js]: If  filepath is null or undefined. HasMagic will throw an error. Maybe I should do something juadgement~

File: src/handleUrl.js (hunk 4 above)
Comment (Genuifx) [inline src/handleUrl.js]: Yep

File: src/handleUrl.js (hunk 1 above)
Comment (SimenB) [inline src/handleUrl.js]: just let it buble all the way up, and catch it at onvocation time, and push the error into `compilation.errors` array.
https://github.com/SimenB/add-asset-html-webpack-plugin/blob/36cd3046cdd9b46171a9fcebdf4835337f5f33b2/src/addAllAssetsToCompilation.js#L50

Comment (Genuifx) [inline src/handleUrl.js]: Got it

File: src/handleUrl.js (hunk 2 above)
Comment (Genuifx) [inline src/handleUrl.js]: `Map` method likes not suitable for this, I use `forEach` instead

File: example/webpack.config.js (hunk 7 above)
Comment (Genuifx) [inline example/webpack.config.js]: not necessary~

Comment (Genuifx) [inline example/webpack.config.js]: so sorry, I just make a mistake. This is necessary for the `npm run example` command to find  files.

File: src/handleUrl.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  assets.forEach(
+    asset =>
+      asset.filepath && globby.hasMagic(asset.filepath)
+        ? globbyAssets.push(asset)
+        : normalAssets.push(asset)
+  );
+  const ret = [];
+  const promises = [];
+  globbyAssets.forEach(asset => {
+    const current = asset;
```
Comment (SimenB) [inline src/handleUrl.js]: `current` is unnecessary

Comment (Genuifx) [inline src/handleUrl.js]: copy that.

File: src/handleUrl.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  const normalAssets = [];
+  // if filepath is null or undefined, just bubble up.
+  assets.forEach(
+    asset =>
+      asset.filepath && globby.hasMagic(asset.filepath)
+        ? globbyAssets.push(asset)
+        : normalAssets.push(asset)
+  );
+  const ret = [];
+  await Promise.all(
```
Comment (SimenB) [inline src/handleUrl.js]: Rewrote this a bit, thoughts?

https://github.com/SimenB/add-asset-html-webpack-plugin/pull/78/commits/4ef0ba202ef35c3aec69babed83a3bba1fb05f79

Comment (Genuifx) [inline src/handleUrl.js]: awesome!

**Comment by SimenB** [comment]
Released as 2.1.0. Thanks!

**Comment by easy-mj** [comment]
image
